src/lib/Util.hs

```diff
@@ -241,3 +242,22 @@ measureSeconds m = do
   ans <- m
   t2 <- liftIO $ getCPUTime
   return (ans, (fromIntegral $ t2 - t1) / 1e12)
+
+-- === zippable class ===
+
+class Zippable f where
+  zipWithZ :: MonadFail m => (a -> b -> m c) -> f a -> f b -> m (f c)
```

Comment:
perhaps call this `zipWithM` since it's a monadic version of `zipWith`?

Reply:
`zipWithM` is already a thing. This is almost a generalization of that though, except that it throws an error if there's a length mismatch. So we could hide `zipWithM` and replace it with ours.